include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1570 @@
+       *            variables whose sensitivities were tracked.
+       *
+       * @return The scalar value of the function.
+       */
+      static ScalarType
+      value(const types::tape_index_type   active_tape_index,
+            const std::vector<ScalarType> &independent_variables);
+
+      /**
+       * Computes the gradient of the scalar field with respect to all
+       * independent variables.
+       *
+       * @param[in] active_tape_index The index of the tape on which the
+       * dependent function is recorded.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] gradient The scalar values of the dependent function's
+       * gradients. It is expected that this vector be of the correct size
+       * (with length <code>n_independent_variables</code>).
+       */
+      static void
+      gradient(const types::tape_index_type   active_tape_index,
+               const std::vector<ScalarType> &independent_variables,
+               Vector<ScalarType> &           gradient);
+
+      /**
+       * Computes the Hessian of the scalar field with respect to all
+       * independent variables.
+       *
+       * @param[in] active_tape_index The index of the tape on which the
+       * dependent function is recorded.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] hessian The scalar values of the dependent function's
+       * Hessian. It is expected that this matrix be of the correct size
+       * (with dimensions
+       * <code>n_independent_variables</code>$\times$<code>n_independent_variables</code>).
+       */
+      static void
+      hessian(const types::tape_index_type   active_tape_index,
+              const std::vector<ScalarType> &independent_variables,
+              FullMatrix<ScalarType> &       hessian);
+
+      //@}
+
+      /**
+       * @name Drivers for vector functions (multiple dependent variables)
+       */
+      //@{
+
+      /**
+       * Computes the values of the vector field.
+       *
+       * @param[in] active_tape_index The index of the tape on which the
+       * dependent function is recorded.
+       * @param[in] n_dependent_variables The number of dependent variables.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] values The scalar values of the dependent functions.
+       * It is expected that this vector be of the correct size
+       * (with length <code>n_dependent_variables</code>).
+       */
+      static void
+      values(const types::tape_index_type   active_tape_index,
+             const unsigned int &           n_dependent_variables,
+             const std::vector<ScalarType> &independent_variables,
```

Comment:
> No need for the reference on the unsigned int.

Sorry, I thought I'd removed all of these. Thanks for spotting this straggler. I've removed its companions too.

> Are you expecting that the number of independent variables is also the length of the array? If so, why pass redundant information?

I think you've misread the names of these two parameters. The first is the expected size of vector that stores the **output** values, and the second is the vector of **input** values. FYI. I send in both `n_dependent_variables` and the result vector `values` because `ADOL-C` doesn't do bounds checking on the destination it writes results to. So this is a last check on the size of this vector before we hand it over to `ADOL-C`.

Reply:
I did misread the names of variables. Thanks for pointing it out!